### PR TITLE
Fix by-component implemented requirement checks

### LINKED ISSUE #770: FedRAMP external constraints validating by-component at wrong layer
### This relates to ...

- [X] the **FedRAMP OSCAL Validations**

### What happened?

In the [FedRAMP OSCAL Documentation](https://automate.fedramp.gov/documentation/ssp/6-security-controls/) it outlines that by-component elements should be at the statements level (control-implementation>implemented-requirements>statements).
Screenshot 2024-10-08 at 4 31 55 PM
We have our OSCAL formatted as outlined in the documentation, but when validating using the enhanced oscal-cli and the fedramp-external-constraints.xml, it flags this as an incorrect structure. It instead gives the following errors, which suggests that these by-component elements should be at the implemented-requirements level rather than statements.

We were hoping you could help us identify whether this is a bug, or a formatting issue with our OSCAL. Here is a snippet of the OSCAL that is causing these validation errors:
```
"implemented-requirements":[
  {
    "uuid":"5b55e601-fa5c-58fc-9596-f8e4ee1cccfc",
    "control-id":"ac-3",
    "props":[
        {
            "name":"control-origination",
            "ns":"https://fedramp.gov/ns/oscal",
            "value":"sp-corporate"
        },
        {
            "name":"control-origination",
            "ns":"https://fedramp.gov/ns/oscal",
            "value":"sp-system"
        },
        {
            "name":"control-origination",
            "ns":"https://fedramp.gov/ns/oscal",
            "value":"customer-configured"
        },
        {
            "name":"implementation-status",
            "ns":"https://fedramp.gov/ns/oscal",
            "value":"implemented"
        }
    ],
    "statements":[
        {
            "statement-id":"ac-3_smt",
            "uuid":"554572df-8a52-54da-803c-170d246f6c3b",
            "by-components":[
                {
                    "component-uuid":"c0e9b4ab-7f2e-54da-9cb3-72894240cc3f",
                    "uuid":"f9086f0c-a65d-597c-9c59-88cf02b30c27",
                    "description":"Private Implementation details and description for the following control statement: AC-03",
                    "implementation-status":{
                        "state":"implemented"
                    },
                    "export":{
                        "provided":[
                            {
                                "uuid":"c39e10b2-28ae-586c-9a2a-93c2983d57c7",
                                "description":"<p>This is what is shared with the customer on export, and what the customer configures<br />\n</p>"
                            }
                        ]
                    }
                }
            ]
        }
    ]
  }
]
```

### Relevant log output

Screenshot 2024-10-08 at 4 41 48 PM

### How do we replicate this issue?

1. Rerun the oscal-cli with an example SSP provided by the reporting user out of band (via email)
2. Locate and confirm if the issue is reproducible or not

### Where, exactly?

- Validations
  - for where exactly a `<by-component/>` SHOULD or MUST be defined in one or implemented requirements
- Documentation on automate.fedramp.gov
- Possible alignment with documentation from NIST at pages.nist.gov/OSCAL, if applicable given confirmation and resolution 

### Other relevant details

Originally posted at https://github.com/metaschema-framework/oscal-cli/issues/55, but GitHub does not permit automatically transferring issues across organizations. I recreated this one manually for @Telos-sa.

### PR DESCRIPTION
**Pull request by aj-stein-gsa**
# Committer Notes

Fixes #770.

- Add new constraint to report against bad practice FedRAMP will not accept for requirements outside of statement.
- Fix current constraint and correct the test, target, and wording.

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- ~If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated?~
- ~If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?~

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

File: src/validations/constraints/content/ssp-all-VALID.xml [hunk 1]
```diff
@@ -212,31 +212,31 @@
       <prop name="control-origination" value="sp-system" ns="https://fedramp.gov/ns/oscal"/>
       <prop name="implementation-status" value="partial" ns="https://fedramp.gov/ns/oscal"/>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/content/ssp-all-VALID.xml:213]: @Rene2mt, I was working on this later last week and I was surprised this bug fix is a little more confusing than I had once thought. I have a question about how we came to have the `prop[@name="implementation-status"]` and not use [the dedicated `implementation-status` assembly](https://pages.nist.gov/OSCAL-Reference/models/v1.1.2/system-security-plan/xml-reference/#/system-security-plan/control-implementation/implemented-requirement/statement/by-component/implementation-status)? I have been gone a while, but this the prop is [also thoroughly documented on our site](https://automate.fedramp.gov/documentation/ssp/6-security-controls/#implementation-status) and the primary `implementation-status` (not the prop) is not documented at all.

Do you know if there was/is a reason for this? It is not immediately related to the PR but tests kept failing as I moved that around and this led me to scratch my head.

File: src/validations/constraints/content/ssp-all-VALID.xml [hunk 2]
```diff
@@ -212,31 +212,31 @@
...
-        </description>
-        <prop ns="https://fedramp.gov/ns/oscal" name="implementation-status" value="implemented"/>
-        <responsible-role role-id="system-admin">
-          <party-uuid>11111111-0000-4000-9000-000000000001</party-uuid>
-        </responsible-role>
-      </by-component>
     </implemented-requirement>
     
     <implemented-requirement uuid="bbbbbbbb-0000-4000-9000-00000000000b" control-id="cm-8">
       <prop name="control-origination" value="sp-system" ns="https://fedramp.gov/ns/oscal"/>
```
Comment (Rene2mt) [inline src/validations/constraints/content/ssp-all-VALID.xml:228]:
```suggestion
      <prop name="control-origination" value="sp-system" ns="https://fedramp.gov/ns/oscal"/>
      <prop ns="https://fedramp.gov/ns/oscal" name="implementation-status" value="partial"/>
```

File: src/validations/constraints/content/ssp-all-VALID.xml [hunk 3]
```diff
@@ -212,31 +212,31 @@
...
       <prop name="implementation-status" value="partial" ns="https://fedramp.gov/ns/oscal"/>
       <statement statement-id="ac-1_stmt.a" uuid="99999999-0000-4000-9000-000000000009">
+        <by-component component-uuid="55555555-0000-4000-9000-000000000005" uuid="aaaaaaaa-0000-4000-9000-00000000000a">
+          <description>
+            <p>Access Control Policy and Procedures (AC-1) is fully implemented in our system.</p>
+          </description>
+          <prop ns="https://fedramp.gov/ns/oscal" name="implementation-status" value="implemented"/>
```
Comment (Rene2mt) [inline src/validations/constraints/content/ssp-all-VALID.xml:219]: Use the `implementations-status` assembly inside the `by-component`, however, it is optional.
```suggestion
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 4]
```diff
@@ -169,7 +169,10 @@
     <context>
         <metapath target="/system-security-plan/control-implementation"/>
         <constraints>
-            <expect id="missing-response-components" target="implemented-requirement" test="count(./by-component) gt 0">
+            <expect id="by-component-requirement-not-inside-statement" target="implemented-requirement" test="not(exists(by-component))">
+                <message>A FedRAMP SSP MUST document only a component-based implemented requirement within a specific statement, not at the control level.</message>
+            </expect>
+            <expect id="missing-response-components" target="implemented-requirement/statement" test="count(./by-component) gt 0">
```
Comment (Rene2mt) [inline src/validations/constraints/fedramp-external-constraints.xml:175]:
```suggestion
            <expect id="by-component-requirement-not-inside-statement" target="implemented-requirement" test="not(exists(by-component))" level="ERROR">
                <message>A FedRAMP SSP MUST document only a component-based implemented requirement within a specific statement, not at the control level.</message>
            </expect>
            <expect id="missing-response-components" target="implemented-requirement/statement" test="count(./by-component) gt 0" level="ERROR">
```

File: src/validations/constraints/content/ssp-by-component-requirement-not-inside-statement-INVALID.xml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    <implemented-requirement uuid="88888888-0000-4000-9000-000000000008" control-id="ac-1">
+      <prop name="control-origination" value="sp-system" ns="https://fedramp.gov/ns/oscal"/>
+      <prop name="implementation-status" value="partial" ns="https://fedramp.gov/ns/oscal"/>
+      <statement statement-id="ac-1_stmt.a" uuid="99999999-0000-4000-9000-000000000009">
+      </statement>
+      <by-component component-uuid="55555555-0000-4000-9000-000000000005" uuid="aaaaaaaa-0000-4000-9000-00000000000a">
+        <description>
+          <p>Access Control Policy and Procedures (AC-1) is fully implemented in our system.</p>
+        </description>
+        <prop ns="https://fedramp.gov/ns/oscal" name="implementation-status" value="implemented"/>
```
Comment (Rene2mt) [inline src/validations/constraints/content/ssp-by-component-requirement-not-inside-statement-INVALID.xml:15]:
```suggestion
```

File: src/validations/constraints/content/ssp-by-component-requirement-not-inside-statement-INVALID.xml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    </implemented-requirement>
+    <implemented-requirement uuid="bbbbbbbb-0000-4000-9000-00000000000b" control-id="cm-8">
+      <prop name="control-origination" value="sp-system" ns="https://fedramp.gov/ns/oscal"/>
+      <statement statement-id="cm-8_stmt.a" uuid="cccccccc-0000-4000-9000-00000000000c">
+      </statement>
+      <by-component component-uuid="55555555-0000-4000-9000-000000000005" uuid="dddddddd-0000-4000-9000-00000000000d">
+        <description>
+          <p>Information System Component Inventory (CM-8) is partially implemented.</p>
+        </description>
+        <prop ns="https://fedramp.gov/ns/oscal" name="implementation-status" value="partial"/>
```
Comment (Rene2mt) [inline src/validations/constraints/content/ssp-by-component-requirement-not-inside-statement-INVALID.xml:29]:
```suggestion
```

**Review by Rene2mt — CHANGES_REQUESTED** [review]
I know this is draft.  Just adding suggestions per the discussion earlier.

**Comment by aj-stein-gsa** [comment]
I am working on integrating Rene's feedback and reviewing the documentation to fix the test. His feedback was very helpful, but we found an unrelated bug when looking into errors with `ssp-all-VALID.xml` that were not making a lot of sense.

**Pull request closed without merging**